Promote Pod Hostname & Subdomain to fields (were annotations)

ArtfulCoder opened the pull request:
Deprecating the podHostName, subdomain and PodHostnames annotations and created corresponding new fields for them on PodSpec and Endpoints types.

Annotation doc: #22564
Annotation code: #20688

ArtfulCoder committed:
Added Hostname and Subdomain field to Pod.Spec

thockin commented on pkg/api/types.go:
```diff
@@ -1225,6 +1225,10 @@ type PodSpec struct {
 	// If specified, these secrets will be passed to individual puller implementations for them to use.  For example,
 	// in the case of docker, only DockerConfig type secrets are honored.
 	ImagePullSecrets []LocalObjectReference `json:"imagePullSecrets,omitempty"`
+	// Specifies the hostname of the Pod
```
. If not provided the pod's hostname will be set to a system-defined value.

ArtfulCoder replied:
done.

thockin commented on pkg/api/types.go:
```diff
@@ -1225,6 +1225,10 @@ type PodSpec struct {
 	ImagePullSecrets []LocalObjectReference `json:"imagePullSecrets,omitempty"`
+	// Specifies the hostname of the Pod
+	Hostname string `json:"hostname,omitempty"`
+	// If specified, the fully qualified Pod hostname will be "<hostname>.<subdomain>.<pod namespace>.svc.<cluster domain>".
```
If not specified, the pod will not have a domainname at all.

ArtfulCoder replied:
done

thockin commented on pkg/api/types.go:
```diff
@@ -1603,6 +1607,10 @@ type Endpoints struct {
+
+	// Its value is the json representation of map[string(IP)][HostRecord]
+	// example: '{"10.245.1.6":{"HostName":"my-webserver"}}'
+	PodHostnames string `json:"podHostnames,omitempty"`
```
This should be an actual field, not a string of a struct.  I think you can instead add an optional hostname field to EndpointAddress, next to IP.

ArtfulCoder replied:
yeah, will do that :)

On Mon, Apr 18, 2016 at 12:02 AM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/api/types.go
> <https://github.com/kubernetes/kubernetes/pull/24362#discussion_r60012366>
> :
>
> > @@ -1603,6 +1607,10 @@ type Endpoints struct {
> >
> >  	// The set of all endpoints is the union of all subsets.
> >  	Subsets []EndpointSubset
> > +
> > +	// Its value is the json representation of map[string(IP)][HostRecord]
> > +	// example: '{"10.245.1.6":{"HostName":"my-webserver"}}'
> > +	PodHostnames string `json:"podHostnames,omitempty"`
>
> This should be an actual field, not a string of a struct. I think you can
> instead add an optional hostname field to EndpointAddress, next to IP.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24362/files/6461c3f70332c1141e578adda35d5e38e728b837#r60012366>
>

ArtfulCoder replied:
done

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -2569,6 +2578,11 @@ func ValidateEndpoints(endpoints *api.Endpoints) field.ErrorList {
 	allErrs = append(allErrs, ValidateEndpointsSpecificAnnotations(endpoints.Annotations, field.NewPath("annotations"))...)
 	allErrs = append(allErrs, validateEndpointSubsets(endpoints.Subsets, field.NewPath("subsets"))...)
+
+	if len(endpoints.PodHostnames) > 0 && !isValidHostnamesMap(endpoints.PodHostnames) {
```
validation of final form is obv. a DNS label

ArtfulCoder replied:
done

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -1345,20 +1345,32 @@ func (kl *Kubelet) GeneratePodHostNameAndDomain(pod *api.Pod) (string, string) {
+	if len(pod.Spec.Hostname) > 0 {
+		if utilvalidation.IsDNS1123Label(pod.Spec.Hostname) {
+			hostname = pod.Spec.Hostname
+		}
```
else log an error ?  Should this proceed or abort the pod run?

thockin commented on pkg/kubelet/kubelet.go:
```diff
@@ -1345,20 +1345,32 @@ func (kl *Kubelet) GeneratePodHostNameAndDomain(pod *api.Pod) (string, string) {
+	if len(pod.Spec.Subdomain) > 0 {
+		if utilvalidation.IsDNS1123Label(pod.Spec.Subdomain) {
+			hostDomain = fmt.Sprintf("%s.%s.svc.%s", pod.Spec.Subdomain, pod.Namespace, clusterDomain)
+		}
```
same - handle errors

thockin commented:
Re-titleing for release note

thockin commented on cluster/addons/dns/skydns-rc.yaml.in:
```diff
@@ -47,7 +47,7 @@ spec:
           mountPath: /var/etcd/data
       - name: kube2sky
-        image: gcr.io/google_containers/kube2sky:1.14
+        image: artfulcoder/kube2sky-amd64:1.16
```
don't forget this

thockin commented on pkg/api/types.go:
```diff
@@ -1626,7 +1632,8 @@ type EndpointAddress struct {
 	// TODO: This should allow hostname or IP, see #4447.
 	IP string
-
+	// Optional: Hostname of this endpoint
```
Add something like "For example, DNS servers can use this."

thockin replied:
still open

ArtfulCoder replied:
done (not pushed)

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -2599,6 +2611,11 @@ func validateEndpointSubsets(subsets []api.EndpointSubset, fldPath *field.Path)
 
 func validateEndpointAddress(address *api.EndpointAddress, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	if len(address.Hostname) > 0 {
```
why is this nested rather than && as above?

thockin replied:
still open

ArtfulCoder replied:
done (not pushed)

thockin commented on pkg/controller/endpoint/endpoints_controller.go:
```diff
@@ -368,6 +372,18 @@ func (e *EndpointController) syncService(key string) {
+				hostRecord := endpoints.HostRecord{
+					HostName: hostname,
+				}
+				podHostNames[string(pod.Status.PodIP)] = hostRecord
```
do we still need the podHostNames map?

thockin replied:
we should stop writing the annotation

thockin replied:
still open

ArtfulCoder replied:
discussed offline. adding todo.

thockin commented on cluster/addons/dns/kube2sky/kube2sky.go:
```diff
@@ -161,28 +161,26 @@ func getSkyMsg(ip string, port int) *skymsg.Service {
 
 func (ks *kube2sky) generateRecordsForHeadlessService(subdomain string, e *kapi.Endpoints, svc *kapi.Service) error {
-	glog.V(4).Infof("Endpoints Annotations: %v", e.Annotations)
+	podHostnames, err := getPodHostnamesFromAnnotation(e.Annotations)
```
TODO: remove this after v1.4 is released and the old annotations are EOL

thockin replied:
yes, 1.4 - that should be long enough?

thockin replied:
still open - I meant to add the TODO

ArtfulCoder replied:
done, not pushed

thockin commented on test/e2e/dns.go:
```diff
@@ -51,12 +51,10 @@ func createDNSPod(namespace, wheezyProbeCmd, jessieProbeCmd string) *api.Pod {
 		ObjectMeta: api.ObjectMeta{
 			Name:      "dns-test-" + string(util.NewUUID()),
 			Namespace: namespace,
-			Annotations: map[string]string{
```
we should keep both cases for now, with a TODO to remove the annotations test case after 1.3

thockin replied:
still open

ArtfulCoder replied:
done

thockin commented:
Also add a TODO at the definitions of the old annotations to remove them after v1.3 is released.

thockin commented:
Mostly pretty minor issues.  Yay!

thockin commented:
We need corresponding docs on this.  Both the DNS related docs and the pod-related docs.

ArtfulCoder commented:
@thockin 
PTAL (last 3 commits. i broke my updates in 3 for easier review)

1. I will rebase after final approval since its a bigger rebase (generated code changed)
2. I will push a skydns image after approval.
3. updated doc and should have addressed all feedback.

thockin commented on cluster/addons/dns/README.md:
```diff
@@ -60,24 +60,55 @@ When enabled, pods are assigned a DNS A record in the form of `pod-ip-address.my
 
 
-####A Records and hostname Based on Pod Annotations - A Beta Feature in Kubernetes v1.2
+####A Records and hostname Based on Pod hostname and Subdomain fields
```
weird capitalization in this title

ArtfulCoder replied:
done (not pushed)

thockin commented:
There are a bunch of comments still unresolved.  Please ack each one when you have done them and push a new commit.

ArtfulCoder commented:
@thockin PTAL

thockin commented on pkg/api/endpoints/util.go:
```diff
@@ -29,6 +29,7 @@ import (
 )
 
 const (
+	// TODO: to be deleted after v1.3 is released
```
Also TODO the HostRecord struct

ArtfulCoder replied:
Done.

thockin commented:
Just a nit.  Make CI pass and you can self-lgtm (unless something huge has to change, in which case ping me).

ArtfulCoder commented:
thanks a lot for the review!

rhcarvalho commented:
@ArtfulCoder sorry if this is in the docs, I couldn't find it.
What would be the behavior when scaling up an RC when the pod template includes the `hostname` field?

The docs talk about a hostname `foo` in a subdomain `bar` can be accessed via `foo.bar...etc`, but what if there are 2+ replicas of the same pod?

ArtfulCoder commented:
@k8s-bot test this please

ArtfulCoder commented:
@k8s-bot test this please

thockin commented:
Great point - docs pls.
On Apr 28, 2016 5:32 AM, "Rodolfo Carvalho" <notifications@github.com>
wrote:

> @ArtfulCoder <https://github.com/ArtfulCoder> sorry if this is in the
> docs, I couldn't find it.
> What would be the behavior when scaling up an RC when the pod template
> includes the hostname field?
>
> The docs talk about a hostname foo in a subdomain bar can be accessed via
> foo.bar...etc, but what if there are 2+ replicas of the same pod?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24362#issuecomment-215380816>
>

rhcarvalho commented:
> Great point - docs pls.

@ArtfulCoder @thockin I didn't find any mention to the case when the number of replicas of a pod >1 in the code that got merged. I expected it somewhere in `cluster/addons/dns/README.md`.

Are the docs coming in a follow up? Could you please share a link? Thanks!

ArtfulCoder commented:
@rhcarvalho , for the case of replicas of a pod > 1, we are planning to introduce petsets, that will have some form of templatization for hostname field. 
@bprashanth to elaborate further since he is working on petsets.

rhcarvalho commented:
@ArtfulCoder thanks for the reply. Currently, is it an error to specify a `hostname` and scale to 2+ replicas? What would the FQDN "foo.bar.my-namespace.svc.cluster.local" point to in the documented example?

I'm following the PetSets PR (https://github.com/kubernetes/kubernetes/pull/18016), though I thought that was focused on persistent volumes.

I'd like to have both scalable persistent storage and pod addressability by hostname to support MongoDB replica sets with persistent storage.

ArtfulCoder commented:
@rhcarvalho If multiple pods have the same hostname, it is not an error. 
You should ideally get back multiple A records. The way I have it right now, causes only one A record to be returned. I will change that behavior to return multiple A records.
